### DISCUSSION TITLE
Separating business logic but public getters not recommended  in Notifiers?

### DISCUSSION THREAD

**Discussion by varunbhalerao56** [rrousselGit/riverpod#2879] [Q&A] [answered]
## Warning Message

>Notifiers should not have public properties/getters. Instead, all their public API should be exposed through the `state` property.

I understand why this is not recommended as any changes in those variables will not be listened to when the state changes.

## The Controller 

<img width="978" alt="image" src="https://github.com/rrousselGit/riverpod/assets/59999892/a1fdb100-590f-45fa-8fcc-f66d7cd0baf2">

## My Explanation

But I feel having this really separates the business logic from the UI? I can easily transfer this business logic elsewhere since the Business Logic is not dependent on the UI for anything other than just calling the functions.

_NOTE: I am aware that this works with something like a text controller but any other value will require a setState after a method call in order to update the UI_

Is there any other way that is recommended over this? Since I would prefer not to use a bad practice all over my apps just cause it works. 

```dart
class ExampleSignUpView extends ConsumerStatefulWidget {
  const ExampleSignUpView({super.key});

  @override
  ConsumerState<ConsumerStatefulWidget> createState() => _ExampleSignUpViewState();
}

class _ExampleSignUpViewState extends ConsumerState<ExampleSignUpView> {

  @override
  void dispose() {
    //? Dispose of all text controllers
  }

  @override
  Widget build(BuildContext context) {
    final controller = ref.watch(_controller.notifier);

    return TextFormField(
      controller: controller.nameController,
      keyboardType: TextInputType.name,
      validator: TextValidators.nameValidator(),
      textInputAction: TextInputAction.next,
      decoration: defaultInputDecorationStyle(labelText: "Name"),
    );
  }
}

```

**Comment by rrousselGit** [comment, marked as answer]
Those controllers should be stored in the notifier's state. You shouldn't have an AsyncNotifier<void>.

Make a class which has a bunch of controllers in it, and have that as state.

**Reply by varunbhalerao56** [reply]
@rrousselGit yes that is what I ended up doing? Just to check you mean something like this? Is there anyway I can improve it or any suggestions that you have?

```dart
part of 'login_view.dart';

/// Controller for [LoginView] handled by [_LoginController] generated by [RiverpodGenerator]
class LoginViewController {
  TextEditingController phoneController;
  TextEditingController otpController;
  Completer<String> onOtpSubmit;
  int? resendToken;
  String? verificationId;
  GlobalKey<FormState> formKey;
  LoginViewController({
    required this.phoneController,
    required this.otpController,
    required this.onOtpSubmit,
    this.resendToken,
    this.verificationId,
    required this.formKey,
  });

  factory LoginViewController.create() => LoginViewController(
        phoneController: TextEditingController(),
        otpController: TextEditingController(),
        onOtpSubmit: Completer<String>(),
        formKey: GlobalKey<FormState>(),
      );
}

@riverpod
class _LoginController extends _$LoginController {
  _LoginController() : super() {
    state = AsyncValue.data(LoginViewController.create());
  }

  @override
  Future<LoginViewController> build() async {
    return state.value ?? LoginViewController.create();
  }

  LoginViewController get _controller => state.value!;

  Future<void> sendOtp(BuildContext context) async {
    state = const AsyncLoading();
    state = await AsyncValue.guard(() async {
      // Sign Up Call
      return Future.value(_controller);
    });
  }

  Future<void> onContinue(BuildContext context) async {
    if (_controller.formKey.currentState?.validate() == false) return;

    toOtp(context);
  }

  void completeOtp(BuildContext context) {
    if (_controller.otpController.text.length != 6) {
      showSnackBar(context, "Invalid OTP");
      return;
    }

    _controller.onOtpSubmit.complete(_controller.otpController.text);
    _controller.onOtpSubmit = Completer<String>();
  }

  void loginViewClear() {
    _controller.phoneController.clear();
    _controller.phoneController.dispose();
  }

  void otpViewClear() {
    _controller.otpController.clear();
    _controller.otpController.dispose();
  }

  void toHome(context) => Navigator.pushAndRemoveUntil(context, LogHomeView.route(), (route) => false);
}

```

**Reply by rrousselGit** [reply]
Kind of, yes.
- I'd remove the "create" factory and move that logic to the Notifier.build method
- Don't initialize state inside the constructor. It's not gonna work anyway
- `build` should always create a new state in your case. Don't use that ??. No need for it
- And use ref.onDispose to destroy your controllers


But generally I wouldn't store controllers in notifiers anyway. These are local widget state. I'd use flutter_hooks for these or a plain StatefulWidget.

**Reply by varunbhalerao56** [reply]
@rrousselGit Aah understood, but what do you think about the separation of business logic and any kind of manipulation? Doesn't this make make it more standalone?

**Reply by rrousselGit** [reply]
Forms are in the UI logic for me, not business one. It's grey area of course

**Reply by varunbhalerao56** [reply]
I initially did this because one of the apps our company was making has a lot of forms and there was a lot of passing the variable around to the action methods like (onSubmit for example). This made things easy but soon realized that this was not the best way haha.  

I am curious though, Are you aware of anything that could replace this and allow me to transfer a bunch UI Logic between multiple projects? I am thinking of hooks to some extent

Anyway, thanks for answering the question!

**Comment by catalunha** [comment]
> But generally I wouldn't store controllers in notifiers anyway. These are local widget state. I'd use flutter_hooks for these or a plain StatefulWidget.

It's very important for me to listen to every issue debate to learn these things. I also agree with you Remi. And I want to learn more and more about Riverpod so I can spread the word.

**Reply by varunbhalerao56** [reply]
What do you think about the separation though? I am totally fine using flutter_hooks but just trying to understand and learn as I go!
